### PR TITLE
[state sync] introduce timeout to consensus requests

### PR DESCRIPTION
**Pull request by phoenix-antigravity**
related to issue https://github.com/libra/libra/issues/3885
now if state sync won't make any progress for configured period of time, it will notify consensus about failed sync request
test is added

### CODE REVIEW

File: state-synchronizer/src/coordinator.rs [hunk 1]
```diff
@@ -382,6 +383,9 @@ impl<T: ExecutorProxyTrait> SyncCoordinator<T> {
         self.check_subscriptions();
         self.peer_manager.remove_requests(local_version);
 
+        if let Some(mut req) = self.sync_request.as_mut() {
+            req.last_progress_tst = SystemTime::now();
```
Comment (zekun000) [inline state-synchronizer/src/coordinator.rs:387]: what is tst?

File: config/src/config/state_sync_config.rs [hunk 2]
```diff
@@ -26,6 +28,7 @@ impl Default for StateSyncConfig {
             long_poll_timeout_ms: 30000,
             max_chunk_limit: 1000,
             max_timeout_ms: 120_000,
+            sync_request_timeout_ms: 60_000,
```
Comment (zekun000) [inline config/src/config/state_sync_config.rs:31]: this is probably too long, I think 10s is enough for a single chunk

**Comment by bmaurer** [comment]
Dumb question: how does the timeout help in the case of #3885? In other words, what was the victim of the issue trying to do that had an infinite timeout? What happens after the timeout occurs that gets them back on track?

What would be the consequence if this timeout fired in a case without misconfiguration

**Comment by zekun000** [comment]
@bmaurer I should have filled more context in the issue, here's what happened

1. one node doesn't upgrade and is on the old-chain with version 5 million.
2. victims upgrade their nodes and are on version 0 but the they share the same keys with the old chain.
3. victims receive valid LedgerInfo about version 5M and try to sync (this is infinite loop until it reaches the target version and blocks consensus) but was unable to finish because execution recognizes it's a fork
4. once we add the timeout, consensus would bail the sync and timeout and start to make progress (although will still stuck for the timeout every time it receives the LedgerInfo from the old chain)

I think it's a good enough fix for now given it's still a problem of forking ourselves, alternatives would be adding a chain id in LedgerInfo which anchors the genesis but won't help if the genesis is the same but the chain forks.

**Comment by bmaurer** [comment]
I see. So in other words, the victim recieves an, apparently valid, LI for version X that is inconsistent with the latest committed LI the victim has. By handing back to consensus we get a chance to recieve consensus messages from other chains / make progress.

In this process, are we able to see that GetChunkResponse is returning an accumulator that is inconsistent with our latest known LI? At least IMHO in production if we were to start downloading a chunk and see a signed LI that conflicts with our accumulator, we'd want to panic / wait for operator intervention

**Comment by zekun000** [comment]
@bmaurer yes, we're able to see the error messages `Fork happens because the current synced_trees doesn't match the txn list provided.` due to our fix to previous sev

Good point, we definitely should make such error much more visible in prod

**Comment by bmaurer** [comment]
That error message seems pretty deep into the codebase... Wondering if there are places closer to where the chunk is validated that would catch this.

Also, if we do want to panic on this in prod... is a timeout the right thing to do?

**Comment by zekun000** [comment]
This is where the chunk being validated I think.
If we do want to panic, we probably should panic directly where the error happens. I wonder if we can make some macro which compile to panic in prod and error in debug mode.

**Pull request closed without merging**